### PR TITLE
Deploy keystone wsgi mod with SSL

### PR DESCRIPTION
**Pull request by nicolasbock**
When using the wsgi apache module we need to enable SSL for it not in
the `keystone.conf` but in apache.

---

This change is [Reviewable](https://reviewable.io/reviews/crowbar/crowbar-openstack/576)

### CODE REVIEW

**Comment by nicolasbock** [comment]
Once non-SSL gating is through, I'll re-trigger with SSL.

**Comment by nicolasbock** [comment]
Keystone is deploying now with SSL. The gating failure is something else I think. I am looking into it right now.

**Comment by nicolasbock** [comment]
The openstack client is failing on `openstack network list`.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/crowbar/crowbar-openstack/576#-:-KVBBGg7w4oZT4l4f-ck:b-5bb0lq)*

**Comment by rsalevsky** [comment]
[cloud-mkcloud7-job-ssl-x86_64](https://trello.com/c/Ex9ZJ8bd/246-cloud-mkcloud7-job-ssl-x86-64)

**Comment by toabctl** [comment]
@nicolasbock Isn't there a ```a2enmod ssl``` missing to be sure ssl is enabled? Also (nitpick), please prefix the commit message with ```keystone: ...```

File: chef/cookbooks/keystone/templates/default/apache_keystone.conf.erb [hunk 1]
```diff
@@ -7,6 +7,14 @@ Listen <%= @bind_service_host %>:<%= @bind_service_port %>
     WSGIPassAuthorization On
     LimitRequestBody 114688
 
+<% if @ssl_enable %>
+    SSLEngine on
+    SSLVerifyDepth 1
```
Comment (toabctl) [inline chef/cookbooks/keystone/templates/default/apache_keystone.conf.erb]: 1 is the default.

File: chef/cookbooks/keystone/recipes/server.rb [hunk 2]
```diff
@@ -166,6 +168,11 @@
       bind_admin_host: bind_admin_host,
       bind_service_port: bind_service_port, # public port
       bind_service_host: bind_service_host,
+      ssl_enable: node[:keystone][:api][:protocol] == "https",
+      ssl_certfile: node[:keystone][:ssl][:certfile],
+      ssl_keyfile: node[:keystone][:ssl][:keyfile],
+      ssl_cert_required: node[:keystone][:ssl][:cert_required],
```
Comment (toabctl) [inline chef/cookbooks/keystone/recipes/server.rb]: you are not using this anywhere. What is the intention of this config parameter?

File: chef/cookbooks/keystone/templates/default/apache_keystone.conf.erb [hunk 3]
```diff
@@ -37,6 +45,14 @@ Listen <%= @bind_admin_host %>:<%= @bind_admin_port %>
...
```
Comment (toabctl) [inline chef/cookbooks/keystone/templates/default/apache_keystone.conf.erb]: same here. 1 is the default

File: chef/cookbooks/keystone/recipes/server.rb (hunk 2 above)
Comment (nicolasbock) [inline chef/cookbooks/keystone/recipes/server.rb]: Hmm, a copy and paste oversight :)

**Comment by nicolasbock** [comment]
@toabctl I have updated the PR

File: chef/cookbooks/keystone/recipes/server.rb (hunk 2 above)
Comment (toabctl) [inline chef/cookbooks/keystone/recipes/server.rb]: If it's no longer used, please remove ```cert_required``` from the schema and from the proposal via a migration.

Comment (nicolasbock) [inline chef/cookbooks/keystone/recipes/server.rb]: We are still using it in `keystone.conf`: https://github.com/nicolasbock/crowbar-openstack/blob/5d92c39b2a4d00bff02bebbe9f564140bb5d6304/chef/cookbooks/keystone/recipes/server.rb#L254

**Comment by nicolasbock** [comment]
@toabctl Could you have another look?

**Comment by toabctl** [comment]
@nicolasbock the ```cert_required``` option is still there. I know it's used in our template but we can remove it there because this is only used in the case where eventlet is used (and because our template is still Mitaka based). And that's no longer possible. So please remove ```cert_required``` with a migration and all places where it's used.
The change from upstream where all this got removed is https://review.openstack.org/#/c/249486/

**Comment by toabctl** [comment]
Alternative is to remove the ```cert_required``` option together with all other eventlet related options when we update the templates. @nicolasbock you can choose :)

**Comment by nicolasbock** [comment]
@toabctl I removed `cert_required`. I am not 100% sure I got the migration right. Could you double check it? Thanks!

**Comment by nicolasbock** [comment]
@toabctl done.

**Comment by nicolasbock** [comment]
@toabctl Could you have another look please?

File: chef/data_bags/crowbar/migrate/keystone/103_remove_cert_required.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+def upgrade(ta, td, a, d)
+  a["ssl"].delete("cert_required") \
+    unless ta["ssl"].key?("cert_required")
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/keystone/103_remove_cert_required.rb]: why not just deleting it?

**Comment by nicolasbock** [comment]
@toabctl I patched a fresh 'instonly' deploy and now the keystone barclamp webform is broken. The SSL option is grayed out and I can't select `https` anymore. I suspect that removing https://github.com/crowbar/crowbar-openstack/pull/576/files#diff-fca7695ab0df9a673d75e40a60e495e9L40 is the problem, but I am not sure how to fix it at this point. More reading rails/haml I guess :)

**Comment by toabctl** [comment]
> @toabctl I patched a fresh 'instonly' deploy and now the keystone barclamp webform is broken. The SSL option is grayed out and I can't select https anymore. I suspect that removing https://github.com/crowbar/crowbar-openstack/pull/576/files#diff-fca7695ab0df9a673d75e40a60e495e9L40 is the problem, but I am not sure how to fix it at this point. More reading rails/haml I guess :)

I think you just did not recompile the assets. your patch works for me

File: chef/cookbooks/keystone/templates/default/apache_keystone.conf.erb [hunk 5]
```diff
@@ -7,6 +7,13 @@ Listen <%= @bind_service_host %>:<%= @bind_service_port %>
     WSGIPassAuthorization On
     LimitRequestBody 114688
 
+<% if @ssl_enable %>
+    SSLEngine on
+    SSLCertificateFile  <%= @ssl_certfile %>
```
Comment (aplanas) [inline chef/cookbooks/keystone/templates/default/apache_keystone.conf.erb]: nit: double space

File: chef/cookbooks/keystone/templates/default/apache_keystone.conf.erb [hunk 6]
```diff
@@ -37,6 +44,13 @@ Listen <%= @bind_admin_host %>:<%= @bind_admin_port %>
...
```
Comment (aplanas) [inline chef/cookbooks/keystone/templates/default/apache_keystone.conf.erb]: nit: double space

**Comment by nicolasbock** [comment]
@toabctl I'll try that again.

**Comment by aplanas** [comment]
Update is not working. If I move from keystone w/out ssl to ssl using this patch, the system try to connect to the http endpoint:

```bash
export OS_AUTH_URL='https://d52-54-77-77-77-01.vo7.cloud.suse.de:5000/v3/'
openstack --debug endpoint list --insecure
```

This produce this:

```
...
REQ: curl -g -i --insecure -X GET https://d52-54-77-77-77-01.vo7.cloud.suse.de:5000/v3/ -H "Accept: application/json" -H "User-Agent: osc-lib keystoneauth1/2.12.2 python-requests/2.11.1 CPython/2.7.9"
...
...
"POST /v3/auth/tokens HTTP/1.1" 201 14549
... "catalog": [{"endpoints": [{"url": "http://10.165.198.2:5000/v3/", "interface": "public", "region": "CustomRegion", "region_id": "CustomRegion", "id": "ef6ecb9179784f569e07a45bf92adb88"}, {"url": "http://d52-54-77-77-77-01.vo7.cloud.suse.de:5000/v3/", "interface": "internal", "region": "CustomRegion", "region_id": "CustomRegion", "id": "0fde6e969aad4c83b14a9b8f6e45f4e3"}, {"url": "http://d52-54-77-77-77-01.vo7.cloud.suse.de:35357/v3/", "interface": "admin", "region": "CustomRegion", "region_id": "CustomRegion", "id": "7977bf0440e64994ade7fbde2bc8bccf"}], "type": "identity", "id": "725a2732f9ca4312bbd35ff3420cbea9", "name": "keystone"},
...
...
Making authentication request to https://d52-54-77-77-77-01.vo7.cloud.suse.de:5000/v3/auth/tokens
...
REQ: curl -g -i --insecure -X GET http://d52-54-77-77-77-01.vo7.cloud.suse.de:35357/v3/ -H "Accept: application/json" -H "User-Agent: osc-lib keystoneauth1/2.12.2 python-requests/2.11.1 CPython/2.7.9"
...
404
```

So everything goes well and use the HTTPS, that generate a good response. But the catalog contain the HTTP URL, so fail after the authentication, when ask for the service.

Is possible to update the catalog before, or the migration from HTTP to HTTPS is not supported / expected?

Also I need to create manually `/etc/keystone/credential-keys` directory

**Comment by nicolasbock** [comment]
@aplanas Thanks for testing! I'll have to check how to update the endpoints during a change. I'll also have to investigate that directory. I can't find any reference to it in the cookbooks.

**Comment by nicolasbock** [comment]
@aplanas I looked a bit into the endpoint challenge. Right now we are bootstrapping keystone the first time we deploy a keystone proposal. If we change the ssl option for keystone, then we will have to add new endpoints and remote the old ones. I tried to simply add a `keystone_register` block to the keystone recipe but that didn't quite work as expected. It adds the new endpoints but it fails at deleting the old ones. After a change the endpoints contain both ssl and non-ssl versions. I think this will require a bit more careful work.

**Comment by toabctl** [comment]
@aplanas , @nicolasbock I think the problem with switching from non-ssl to ssl is unrelated to this PR. I'm pretty sure this was broken before. I would ignore it (create an extra bug report for it) and check the case if a fresh deployment with ssl works or not.

**Comment by nicolasbock** [comment]
@toabctl Yes I agree. I don't think that it's working for a fresh SSL deployment either though. Neutron is not starting up. I am looking into it.

**Comment by toabctl** [comment]
The neutron thing is another problem. Can we merge this, @aplanas , @nicolasbock ? it is solving the mentioned bug...

**Comment by toabctl** [comment]
I filled bsc#1011199 for the followup problem

**Pull request merged**